### ISSUE TITLE
When running the `static-xor-mlp` test, loss approaches zero, then jumps abruptly

### ISSUE THREAD

**Issue by jcberentsen** [hasktorch/hasktorch#301]
In the output below the loss seems to converge to 8.9571e-21 then it inexplicably jumps to 0.5117

Is the instability to be expected from this simple example?

```
$ stack run static-xor-mlp
Tensor Float []  8.2879e-7
Tensor Float []  1.8830e-7
Tensor Float []  6.2519e-8
Tensor Float []  2.1554e-8
Tensor Float []  6.1962e-9
Tensor Float []  1.8083e-9
Tensor Float []  4.9934e-10
Tensor Float []  1.5774e-10
Tensor Float []  4.6754e-11
Tensor Float []  1.3682e-11
Tensor Float []  3.8478e-12
Tensor Float []  1.0429e-12
Tensor Float []  2.9771e-13
Tensor Float []  8.2055e-14
Tensor Float []  2.6305e-14
Tensor Float []  6.9057e-15
Tensor Float []  9.8372e-16
Tensor Float []  2.3195e-16
Tensor Float []  6.5176e-17
Tensor Float []  1.7209e-17
Tensor Float []  4.8622e-18
Tensor Float []  1.4925e-18
Tensor Float []  4.5743e-19
Tensor Float []  1.0700e-19
Tensor Float []  3.1411e-20
Tensor Float []  8.9571e-21
Tensor Float []  0.5117   
Tensor Float []  0.4805   
Tensor Float []  0.5156   
Tensor Float []  0.4844   
Tensor Float []  0.4883   
Tensor Float []  0.4805   
Tensor Float []  0.5547   
Tensor Float []  0.4414   
Tensor Float []  0.5039   
Tensor Float []  0.2891   
Tensor Float []  0.2539   
Tensor Float []  0.1216   
Tensor Float []  0.1304   
Tensor Float []  0.1301   
MLP {layer0 = Linear {linearWeight = UnsafeMkParameter (IndependentTensor {toDependent = Tensor Float [4,2] [[-7.3322   , -6.6333   ],
                    [ 1.3172   ,  0.6941   ],
                    [-13.6337   , -16.1908   ],
                    [ 4.0229   , -6.1787   ]]}), linearBias = UnsafeMkParameter (IndependentTensor {toDependent = Tensor Float [4] [-3.2960   ,  0.8008   ,  2.8894   , -1.7450   ]})}, layer1 = Linear {linearWeight = UnsafeMkParameter (IndependentTensor {toDependent = Tensor Float [4,4] [[ 0.8016   ,  3.9377   ,  2.5386   , -8.0361   ],
                    [-3.0556   , -4.0443   , -0.6912   ,  11.8981   ],
                    [ 10.1147   ,  3.1606   ,  11.7324   , -1.5009   ],
                    [ 2.1492   ,  3.8292   ,  2.8993   , -7.3043   ]]}), linearBias = UnsafeMkParameter (IndependentTensor {toDependent = Tensor Float [4] [ 5.1598   , -5.7846   ,  3.5713   ,  5.0281   ]})}, layer2 = Linear {linearWeight = UnsafeMkParameter (IndependentTensor {toDependent = Tensor Float [1,4] [[-0.5687   ,  3.8690   , -7.5810   , -3.3823   ]]}), linearBias = UnsafeMkParameter (IndependentTensor {toDependent = Tensor Float [1] [ 0.1720   ]})}}
```

This was ran on a macBook pro on commit `dc0d33974319dc5d18a6b9d760b16bdc281298d6`

**Comment by junjihashimoto** [comment]
@jcberentsen 
Thank you for reporting it.
I think we do not expect such instability of the loss. (The sample uses adam.)

**Comment by austinvhuang** [comment]
I just ran it a couple times and didn't observe that behavior. Probably just luck of the RNG + a slightly aggressive learning rate. Could submit a PR to tweak the learning rate to mitigate this or do some detailed side-by-side comparisons with an equivalently parameterized pytorch model, but I think it's an unlucky RNG draw.

```
Iteration: 100 | Loss: Tensor Float []  0.3401   
Iteration: 200 | Loss: Tensor Float []  0.1926   
Iteration: 300 | Loss: Tensor Float []  3.4295e-2
Iteration: 400 | Loss: Tensor Float []  4.4569e-3
Iteration: 500 | Loss: Tensor Float []  1.0748e-5
Iteration: 600 | Loss: Tensor Float []  1.9814e-7
Iteration: 700 | Loss: Tensor Float []  2.3710e-8
Iteration: 800 | Loss: Tensor Float []  1.8468e-9
Iteration: 900 | Loss: Tensor Float []  4.6363e-12
Iteration: 1000 | Loss: Tensor Float []  7.1054e-14
Iteration: 1100 | Loss: Tensor Float []  1.7764e-14
Iteration: 1200 | Loss: Tensor Float []  3.1974e-14
Iteration: 1300 | Loss: Tensor Float []  6.0396e-14
Iteration: 1400 | Loss: Tensor Float []  2.8422e-14
Iteration: 1500 | Loss: Tensor Float []  3.1974e-14
Iteration: 1600 | Loss: Tensor Float []  3.0198e-14
Iteration: 1700 | Loss: Tensor Float []  5.6843e-14
Iteration: 1800 | Loss: Tensor Float []  3.5527e-14
Iteration: 1900 | Loss: Tensor Float []  3.5527e-14
Iteration: 2000 | Loss: Tensor Float []  1.7764e-14
Final Model:
0, 0 => Tensor Float []  5.9605e-8
0, 1 => Tensor Float []  1.0000   
1, 0 => Tensor Float []  1.0000   
1, 1 => Tensor Float []  1.7881e-7
(base) austinvh@penguin:~/hasktorch$ stack run xor-mlp
Iteration: 100 | Loss: Tensor Float []  0.2236   
Iteration: 200 | Loss: Tensor Float []  0.2780   
Iteration: 300 | Loss: Tensor Float []  0.3861   
Iteration: 400 | Loss: Tensor Float []  1.8975e-2
Iteration: 500 | Loss: Tensor Float []  8.4227e-4
Iteration: 600 | Loss: Tensor Float []  2.9578e-6
Iteration: 700 | Loss: Tensor Float []  6.7611e-8
Iteration: 800 | Loss: Tensor Float []  1.3037e-9
Iteration: 900 | Loss: Tensor Float []  6.2812e-12
Iteration: 1000 | Loss: Tensor Float []  1.4211e-14
Iteration: 1100 | Loss: Tensor Float []  7.1054e-14
Iteration: 1200 | Loss: Tensor Float []  7.2831e-14
Iteration: 1300 | Loss: Tensor Float []  8.8818e-14
Iteration: 1400 | Loss: Tensor Float []  5.6843e-14
Iteration: 1500 | Loss: Tensor Float []  4.4409e-14
Iteration: 1600 | Loss: Tensor Float []  5.6843e-14
Iteration: 1700 | Loss: Tensor Float []  7.2831e-14
Iteration: 1800 | Loss: Tensor Float []  8.8818e-14
Iteration: 1900 | Loss: Tensor Float []  4.4409e-14
Iteration: 2000 | Loss: Tensor Float []  5.1514e-14
Final Model:
0, 0 => Tensor Float []  2.3842e-7
0, 1 => Tensor Float []  1.0000   
1, 0 => Tensor Float []  1.0000   
1, 1 => Tensor Float []  1.1921e-7
(base) austinvh@penguin:~/hasktorch$ stack run xor-mlp
Iteration: 100 | Loss: Tensor Float []  0.4830   
Iteration: 200 | Loss: Tensor Float []  1.3863e-3
Iteration: 300 | Loss: Tensor Float []  2.5932e-4
Iteration: 400 | Loss: Tensor Float []  9.8830e-6
Iteration: 500 | Loss: Tensor Float []  2.9397e-8
Iteration: 600 | Loss: Tensor Float []  1.0708e-9
Iteration: 700 | Loss: Tensor Float []  4.5986e-11
Iteration: 800 | Loss: Tensor Float []  3.9257e-13
Iteration: 900 | Loss: Tensor Float []  5.1514e-14
Iteration: 1000 | Loss: Tensor Float []  3.1974e-14
Iteration: 1100 | Loss: Tensor Float []  8.8818e-15
Iteration: 1200 | Loss: Tensor Float []  1.7764e-15
Iteration: 1300 | Loss: Tensor Float []  7.1054e-15
Iteration: 1400 | Loss: Tensor Float []  7.1054e-15
Iteration: 1500 | Loss: Tensor Float []  7.1054e-15
Iteration: 1600 | Loss: Tensor Float []  1.4211e-14
Iteration: 1700 | Loss: Tensor Float []  7.1054e-15
Iteration: 1800 | Loss: Tensor Float []  8.8818e-15
Iteration: 1900 | Loss: Tensor Float []  1.4211e-14
Iteration: 2000 | Loss: Tensor Float []  1.4211e-14
Final Model:
0, 0 => Tensor Float []  0.0000
0, 1 => Tensor Float []  1.0000   
1, 0 => Tensor Float []  1.0000   
1, 1 => Tensor Float []  5.9605e-8
```

**Comment by austinvhuang** [comment]
I just ran it and see some jumping around but nothing quite that large, but it could be an unlucky RNG draw + the learning rate being a little high.

Not able to edit now but can probably tweak this example's parameters to be closer to the non-static version. The number of iterations here is also an order of magnitude more than necessary and in my past runs the untyped version converges pretty consistently from what I've seen.

**Comment by jcberentsen** [comment]
Yes, I noticed learning rate seemed a bit big, maybe expecting too much of Adam? I did a few runs and saw the problem several times (On CPU)

Maybe the learning can be stopped when the |loss| converges near 0? (maybe 1E-9), or simpler as you mention, limit the number of iterations.

I think a reason for the jumping would be enlightening. As a user I would worry if the issue is with the learning code, or hasktorch.

**Comment by austinvhuang** [comment]
@jcberentsen I believe this has been reported elsewhere with adam, for example:

https://discuss.pytorch.org/t/loss-suddenly-increases-using-adam-optimizer/11338
https://datascience.stackexchange.com/questions/25024/strange-behavior-with-adam-optimizer-when-training-for-too-long

That said, probably can tweak the parameters in this example so users are less prone to running into it.
